Kernel/Ptrace.cpp

```diff
@@ -0,0 +1,153 @@
+#include <Kernel/Process.h>
+#include <Kernel/Ptrace.h>
+#include <Kernel/Thread.h>
+#include <Kernel/ThreadTracer.h>
+#include <Kernel/VM/MemoryManager.h>
+#include <Kernel/VM/ProcessPagingScope.h>
+
+namespace Ptrace {
+
+KResultOr<u32> handle_syscall(const Kernel::Syscall::SC_ptrace_params& params, Process& caller)
+{
+    if (params.request == PT_TRACE_ME) {
+        if (Thread::current->tracer())
+            return KResult(-EBUSY);
+
+        caller.set_wait_for_tracer_at_next_execve(true);
+        return KSuccess;
+    }
+
+    if (params.pid == caller.pid())
+        return KResult(-EINVAL);
+
+    Thread* peer = nullptr;
+    {
+        InterruptDisabler disabler;
+        peer = Thread::from_tid(params.pid);
+    }
+    if (!peer)
+        return KResult(-ESRCH);
+
+    if ((peer->process().uid() != caller.euid())
+        || (peer->process().uid() != peer->process().euid())) // Disallow tracing setuid processes
+        return KResult(-EACCES);
+
+    if (params.request == PT_ATTACH) {
+        if (peer->tracer()) {
+            return KResult(-EBUSY);
+        }
+        peer->start_tracing_from(caller.pid());
+        if (peer->state() != Thread::State::Stopped && !(peer->has_blocker() && peer->blocker().is_reason_signal()))
+            peer->send_signal(SIGSTOP, &caller);
+        return KSuccess;
+    }
+
+    auto* tracer = peer->tracer();
+
+    if (!tracer)
+        return KResult(-EPERM);
+
+    if (tracer->tracer_pid() != caller.pid())
+        return KResult(-EBUSY);
+
+    if (peer->state() == Thread::State::Running)
+        return KResult(-EBUSY);
+
+    switch (params.request) {
+    case PT_CONTINUE:
+        peer->send_signal(SIGCONT, &caller);
+        break;
+
+    case PT_DETACH:
+        peer->stop_tracing();
+        peer->send_signal(SIGCONT, &caller);
+        break;
+
+    case PT_SYSCALL:
+        tracer->set_trace_syscalls(true);
+        peer->send_signal(SIGCONT, &caller);
+        break;
+
+    case PT_GETREGS: {
+        if (!tracer->has_regs())
+            return KResult(-EINVAL);
+
+        PtraceRegisters* regs = reinterpret_cast<PtraceRegisters*>(params.addr);
+        if (!caller.validate_write(regs, sizeof(PtraceRegisters)))
+            return KResult(-EFAULT);
+
+        {
+            SmapDisabler disabler;
+            *regs = tracer->regs();
+        }
+        break;
+    }
+
+    case PT_SETREGS: {
+        if (!tracer->has_regs())
+            return KResult(-EINVAL);
+
+        if (!caller.validate_read(params.addr, sizeof(PtraceRegisters)))
+            return KResult(-EFAULT);
+
+        {
+            SmapDisabler disabler;
+            PtraceRegisters* regs = reinterpret_cast<PtraceRegisters*>(params.addr);
+            RegisterState* peer_saved_registers = reinterpret_cast<RegisterState*>(peer->kernel_stack_top() - sizeof(RegisterState));
```

Comment:
The somewhat awkwardly named `RegisterState& Thread::get_register_dump_from_stack()` can get you a reference to this place :)